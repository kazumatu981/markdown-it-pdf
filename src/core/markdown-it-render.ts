import MarkdownIt from 'markdown-it';
import fsPromises from 'fs/promises';

export interface Styles {
    internalUrls: string[];
    externalUrls: string[];
}

export class MarkdownItRender {
    private _md: MarkdownIt = new MarkdownIt();
    private _styles: Styles = {
        internalUrls: [],
        externalUrls: [],
    };

    /**
     * Adds the given URLs to the list of internal styles for this instance.
     *
     * @param {string[]} urls - The URLs of the styles to add.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addStyles(urls: string[]): this {
        this._styles.internalUrls = urls;
        return this;
    }

    /**
     * Sets the external style URLs of this instance.
     *
     * @param {string[]} urls - The URLs to set as external styles.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addExternalStyles(urls: string[]): this {
        this._styles.externalUrls = urls;
        return this;
    }

    /**
     * Uses the provided plugin with optional parameters.
     *
     * @param {MarkdownIt.PluginWithParams} plugin - The plugin to use.
     * @param {any[]} params - Optional parameters for the plugin.
     * @return {this} The MarkdownItRender instance for chaining.
     */
    public use(plugin: MarkdownIt.PluginWithParams, ...params: any[]): this {
        this._md.use(plugin, ...params);
        return this;
    }

    /**
     * Renders the given markdown string into HTML.
     *
     * @param {string} markdown - The markdown string to render.
     * @return {string} The rendered HTML string.
     */
    public render(markdown: string): string {
        const htmlBody = this._md.render(markdown);
        const styleTags = MarkdownItRender.generateStyleTags(
            this._styles.internalUrls,
            this._styles.externalUrls
        );
        return MarkdownItRender.htmlTemplate
            .replace(MarkdownItRender.styles, styleTags)
            .replace(MarkdownItRender.contents, htmlBody);
    }
    /**
     * Asynchronously renders a markdown file to HTML.
     *
     * @param {string} markdownFilePath - The path to the markdown file.
     * @return {Promise<string>} A promise that resolves to the rendered HTML.
     */
    public async renderFromFileAsync(
        markdownFilePath: string
    ): Promise<string> {
        const markdown = await fsPromises.readFile(markdownFilePath, 'utf8');
        return this.render(markdown);
    }

    //#region private members
    //#region Static Variables
    private static styles = '<!-- [[[STYLES]]] -->';
    private static contents = '<!-- [[[CONTENTS]]] -->';

    private static htmlTemplate = `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    ${MarkdownItRender.styles}
</head>
<body>
    ${MarkdownItRender.contents}
</body>`;
    //#endregion

    //#region Static Methods
    /**
     * Generates style tag based on the given URL.
     * @param {string} stylePath css file path
     * @returns {string} generated style tag
     */
    private static generateStyleTag(stylePath: string): string {
        return `<link rel="stylesheet" type="text/css" href="${stylePath}">`;
    }
    /**
     * Generates style tags based on the given internal and external URLs.
     *
     * @param {string[]} internalUrls - An optional array of internal URLs.
     * @param {string[]} externalUrls - An optional array of external URLs.
     * @return {string} The generated style tags as a string.
     */
    private static generateStyleTags(
        internalUrls?: string[],
        externalUrls?: string[]
    ): string {
        return (
            (internalUrls?.map(MarkdownItRender.generateStyleTag).join('\n') ??
                '') +
            (externalUrls?.map(MarkdownItRender.generateStyleTag).join('\n') ??
                '')
        );
    }
    //#endregion
    //#endregion
}
