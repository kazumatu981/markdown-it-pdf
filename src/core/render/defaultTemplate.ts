export const defaultTemplateSource = `
<!doctype html>
<html>
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta http-equiv="X-UA-Compatible" content="ie=edge" />
        <title>Document</title>
        {{#hljs}}
        <script src="{{js}}"></script>
        <link rel="stylesheet" type="text/css" href="{{css}}" />
        {{/hljs}}
        {{#styles}}
        <link rel="stylesheet" type="text/css" href="{{.}}" />
        {{/styles}}
    </head>
    <body>
        {{#hljs}}
        <script>hljs.highlightAll();</script>
        {{/hljs}}
        {{{body}}}
    </body>
</html>
`;
