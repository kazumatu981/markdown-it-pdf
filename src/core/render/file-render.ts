import fsPromises from 'fs/promises';

export interface FileRender {
    renderFromFile(filePath: string): Promise<string | Buffer>;
}

export class PlainTextRender implements FileRender {
    private _charset: string;
    constructor(charset: string) {
        this._charset = charset;
    }
    renderFromFile(filePath: string): Promise<string> {
        return fsPromises.readFile(filePath, {
            encoding: this._charset as BufferEncoding,
        });
    }
}

export class BinaryRender implements FileRender {
    renderFromFile(filePath: string): Promise<Buffer> {
        return fsPromises.readFile(filePath);
    }
}

export const utf8PlainTextRender = new PlainTextRender('utf8');
export const binaryRender = new BinaryRender();
export const defaultRender = binaryRender;
