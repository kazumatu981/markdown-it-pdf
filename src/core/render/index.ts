export * from './file-render';
export * from './markdown-it-render';
