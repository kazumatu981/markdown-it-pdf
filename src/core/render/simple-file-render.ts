import fsPromises from 'fs/promises';
import type { FileRender } from './file-render';

/**
 * The implementation of the render function for plain text or binary files.
 */
export class SimpleFileRender implements FileRender {
    _charset?: string;

    /**
     * constructs a new instance of the SimpleFileRender class.
     * @param charset - The charset of the file.
     */
    public constructor(charset?: string) {
        this._charset = charset;
    }

    /**
     * Reads the contents of a file and returns it as a string or Buffer.
     * @param filePath {string} - The path of the file to read.
     * @returns {Promise<string | Buffer>} A promise that resolves to the contents of the file as a string or Buffer.
     */
    public renderFromFile(filePath: string): Promise<string | Buffer> {
        return fsPromises.readFile(filePath, {
            encoding: this._charset as BufferEncoding,
        });
    }
}

/**
 * The plain text file render function.
 */
export const utf8PlainTextRender = new SimpleFileRender('utf8');
/**
 * The binary file render function.
 */
export const binaryRender = new SimpleFileRender();
/**
 * The default render function. render as binary file.
 */
export const defaultRender = binaryRender;
