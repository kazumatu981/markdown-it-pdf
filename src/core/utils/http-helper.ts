import http from 'http';
import { type Logger } from '../../common';

//#region constants
const privatePortRange = {
    min: 49152,
    max: 65535,
};

const defaultListeningOptions = {
    retry: 10,
    range: privatePortRange,
};
//#endregion

//#region types and interfaces

/**
 * The server that is listening on a given port
 */
interface ListeningHttpServer {
    /**
     * The port that the server is listening on.
     */
    port: number;
    /**
     * The http server that is listening on the specified port.
     */
    httpServer: http.Server;
}

/**
 * The range of ports to search for
 */
export interface Range {
    /**
     * The minimum port to search for
     */
    min?: number;
    /**
     * The maximum port to search for
     */
    max?: number;
}

/**
 * Options for listening on a port
 */
export interface ListeningOptions {
    /**
     * The number of times to retry if the port is in use
     */
    retry?: number;
    /**
     * The range of ports to search for
     */
    range?: Range;
}

//#endregion

//#region functions

//#region exported functions
/**
 * Tries to listen on a given port. If the port is not specified, it tries to
 * listen on a random port within a specified range. If the specified port is
 * in use, it retries the specified number of times.
 *
 * @param {number | undefined} port The port to listen on. If not specified, it will try to listen on
 * a random port within the specified range.
 * @param {ListeningOptions | undefined} options The options for listening on a port.
 * @param {Logger | undefined} logger The logger to log the events.
 * @returns {Promise<ServerPort | undefined>} A promise that resolves to the port that the server is listening on
 * or undefined if the specified port is in use and cannot be listened on.
 */
export async function tryToListen(
    port?: number,
    options?: ListeningOptions,
    logger?: Logger
): Promise<ListeningHttpServer | undefined> {
    let listeningHttpServer = undefined;
    if (port != undefined) {
        logger?.debug('Try to Listen on port %d', port);
        listeningHttpServer = await tryToListenCore(port, logger);
    } else {
        logger?.debug('Try to Listen on port random mode');
        for (
            let i = 0;
            i < (options?.retry ?? defaultListeningOptions.retry);
            i++
        ) {
            port = getRandom(options?.range);
            listeningHttpServer = await tryToListenCore(port, logger);
            if (listeningHttpServer !== undefined) {
                break; // success
            }
        }
    }
    return listeningHttpServer;
}
//#endregion

//#region internal functions
/**
 * Generates a random number within a specified range.
 *
 * @param {Range | undefined} range - The range within which the random number
 * should be generated. If not specified, the function will use the default range
 * of 49152 to 65535.
 * @returns {number} A random number within the specified range.
 * @throws {Error} If the range is invalid (min > max).
 */
function getRandom(range?: Range): number {
    // Set the default range if not specified
    const candidate = {
        min: range?.min ?? defaultListeningOptions.range.min,
        max: range?.max ?? defaultListeningOptions.range.max,
    };

    // Check if the range is valid
    if (candidate.min > candidate.max) {
        throw new Error('Invalid range');
    }

    // Generate a random number within the range
    return Math.floor(
        Math.random() * (candidate.max - candidate.min) + candidate.min
    );
}

/**
 * Tries to listen on a given port. If the port is in use, it rejects the
 * promise.
 *
 * @param {number} port The port to listen on.
 * @param {Logger | undefined} logger The logger to log the events.
 * @returns {Promise<ServerPort | undefined>} A promise that resolves to the port that the server is listening on
 * or undefined if the specified port is in use and cannot be listened on.
 */
function tryToListenCore(
    port: number,
    logger?: Logger
): Promise<ListeningHttpServer | undefined> {
    // Create a new http server
    const server = http.createServer();

    // Create a promise that resolves to the port that the server is listening on
    // or undefined if the specified port is in use and cannot be listened on
    return new Promise((resolve) => {
        // Start listening on the specified port
        server.listen(port, () => {
            // Log the reserved port
            logger?.debug('reserved port: %d', port);
            // Resolve the promise with the port and the server
            resolve({
                port,
                httpServer: server,
            });
        });
        // If there is an error while listening on the port, log the error and
        // resolve the promise with undefined
        server.on('error', (error) => {
            logger?.debug('listen error: %s', error.message);
            logger?.trace(error.stack);
            logger?.warn('port %d may be used.', port);
            resolve(undefined);
        });
    });
}
//#endregion

//#endregion
