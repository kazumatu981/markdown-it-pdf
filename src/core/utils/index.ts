export * from './http-helper';
export * from './path-resolver';
