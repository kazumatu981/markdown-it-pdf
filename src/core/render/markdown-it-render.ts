import MarkdownIt from 'markdown-it';
import fsPromises from 'fs/promises';
import { Logger } from '../../common/logger';
import { type Styles } from '../../common/configure';
import { type FileRender } from './file-render';

export class MarkdownItRender extends MarkdownIt implements FileRender {
    public _logger?: Logger;

    private _styles: Styles = {
        internalUrls: [],
        externalUrls: [],
    };

    /**
     * Adds the given URLs to the list of internal styles for this instance.
     *
     * @param {string[]} urls - The URLs of the styles to add.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addStyles(urls: string[]): this {
        this._styles.internalUrls.push(...urls);
        return this;
    }

    /**
     * Sets the external style URLs of this instance.
     *
     * @param {string[]} urls - The URLs to set as external styles.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addExternalStyles(urls: string[]): this {
        this._styles.externalUrls.push(...urls);
        return this;
    }

    /**
     * Renders the given markdown string into HTML.
     *
     * @param {string} markdown - The markdown string to render.
     * @return {string} The rendered HTML string.
     */
    public render(markdown: string): string {
        this._logger?.debug(`render() called.`);
        this._logger?.debug(`styles: %o`, this._styles);
        const htmlBody = super.render(markdown);
        const styleTags = MarkdownItRender.generateStyleTags(
            this._styles.internalUrls,
            this._styles.externalUrls
        );
        return MarkdownItRender.htmlTemplate
            .replace(MarkdownItRender.styles, styleTags)
            .replace(MarkdownItRender.contents, htmlBody);
    }
    /**
     * Asynchronously renders a markdown file to HTML.
     *
     * @param {string} markdownFilePath - The path to the markdown file.
     * @return {Promise<string>} A promise that resolves to the rendered HTML.
     */
    public async renderFromFile(markdownFilePath: string): Promise<string> {
        const markdown = await fsPromises.readFile(markdownFilePath, 'utf8');
        return this.render(markdown);
    }

    //#region private members
    //#region Static Variables
    private static styles = '<!-- [[[STYLES]]] -->';
    private static contents = '<!-- [[[CONTENTS]]] -->';

    private static htmlTemplate = `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    ${MarkdownItRender.styles}
</head>
<body>
    ${MarkdownItRender.contents}
</body>`;
    //#endregion

    //#region Static Methods
    /**
     * Generates style tag based on the given URL.
     * @param {string} stylePath css file path
     * @returns {string} generated style tag
     */
    private static generateStyleTag(stylePath: string): string {
        return `<link rel="stylesheet" type="text/css" href="${stylePath}">`;
    }
    /**
     * Generates style tags based on the given internal and external URLs.
     *
     * @param {string[]} internalUrls - An optional array of internal URLs.
     * @param {string[]} externalUrls - An optional array of external URLs.
     * @return {string} The generated style tags as a string.
     */
    private static generateStyleTags(
        internalUrls: string[],
        externalUrls: string[]
    ): string {
        return (
            internalUrls.map(MarkdownItRender.generateStyleTag).join('\n') +
            '\n' +
            externalUrls.map(MarkdownItRender.generateStyleTag).join('\n')
        );
    }
    //#endregion
    //#endregion
}
