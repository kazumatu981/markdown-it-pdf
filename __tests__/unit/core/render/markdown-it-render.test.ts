import { jest, describe, it, expect } from '@jest/globals';
import { MarkdownItRender } from '../../../../src/core/render/markdown-it-render';

import { mockLogger } from '../../../utils/mock-logger';

import fsAsync from 'fs/promises';
import { PathLike } from 'fs';

const MarkdownItSup = require('markdown-it-sup');

jest.mock('fs/promises');
const templateContents = `
<html>
    <head>
        {{#{styles}}}
        <link rel="stylesheet" type="text/css" href="{{.}}" />
        {{/styles}}
    </head>
    <body>
        {{{body}}}
    </body>
`;

// TODO Add test about Hljs configure.

describe('CoreLibrary Unit Tests - MarkdownItRender', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });
    it('Basic Render Test: Can render from file', async () => {
        (
            fsAsync.readFile as jest.MockedFunction<typeof fsAsync.readFile>
        ).mockResolvedValueOnce('# test\n\nhello world');
        const markdownItRender = new MarkdownItRender();
        const result = await markdownItRender.renderFromFile('./test.md');
        expect(result).toMatchSnapshot();
    });
    it('Basic Render Test: Can render from string', async () => {
        const markdownItRender = new MarkdownItRender();
        await markdownItRender.configureTemplate();
        const result = markdownItRender.render('# test\n\nhello world');
        expect(result).toMatchSnapshot();
    });
    it('Render Test: Can use plugins', async () => {
        const markdownItRender = new MarkdownItRender();
        markdownItRender.use(MarkdownItSup);
        await markdownItRender.configureTemplate();
        const result = markdownItRender.render('# test\n\nhello ^world^');
        expect(result).toMatchSnapshot();
    });
    it('Render Test: Styles are rendered (styleFilePaths)', async () => {
        const markdownItRender = new MarkdownItRender();
        markdownItRender.addStyles(['./test.css']);
        const result = markdownItRender.render('# test\n\nhello world');
        await markdownItRender.configureTemplate();
        expect(result).toMatchSnapshot();
    });
    it('Render Test: Styles are rendered (externalStylesUrls)', async () => {
        const markdownItRender = new MarkdownItRender();
        markdownItRender.addExternalStyles(['https://hoo.bar/styles/test.css']);
        const result = markdownItRender.render('# test\n\nhello world');
        expect(result).toMatchSnapshot();
    });
    it('Render Test: with logger', async () => {
        const markdownItRender = new MarkdownItRender();
        markdownItRender._logger = mockLogger;
        markdownItRender.addStyles(['./test.css']);
        markdownItRender.addExternalStyles(['https://hoo.bar/styles/test.css']);
        const result = markdownItRender.render('# test\n\nhello world');
        expect(mockLogger.debug).toMatchSnapshot();
    });

    it('readTemplateFromFile', async () => {
        const markdownItRender = new MarkdownItRender();
        (
            fsAsync.readFile as jest.MockedFunction<typeof fsAsync.readFile>
        ).mockImplementation(((filePath: any, options: any) => {
            if (filePath === './template.html') {
                return Promise.resolve(templateContents);
            } else if (filePath === './test.md') {
                return Promise.resolve('# test\n\nhello world');
            }
        }) as any);

        const result = await markdownItRender.configureTemplate({
            templatePath: './template.html',
        });

        expect(result['templateSource']).toMatchSnapshot();
    });
});
