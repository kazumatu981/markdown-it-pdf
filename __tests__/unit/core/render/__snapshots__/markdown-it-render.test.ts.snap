// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreLibrary Unit Tests - MarkdownItRender Basic Render Test: Can render from file 1`] = `
"
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    

</head>
<body>
    <h1>test</h1>
<p>hello world</p>

</body>"
`;

exports[`CoreLibrary Unit Tests - MarkdownItRender Basic Render Test: Can render from string 1`] = `
"
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    

</head>
<body>
    <h1>test</h1>
<p>hello world</p>

</body>"
`;

exports[`CoreLibrary Unit Tests - MarkdownItRender Render Test: Can use plugins 1`] = `
"
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    

</head>
<body>
    <h1>test</h1>
<p>hello <sup>world</sup></p>

</body>"
`;

exports[`CoreLibrary Unit Tests - MarkdownItRender Render Test: Styles are rendered (externalStylesUrls) 1`] = `
"
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    
<link rel="stylesheet" type="text/css" href="https://hoo.bar/styles/test.css">
</head>
<body>
    <h1>test</h1>
<p>hello world</p>

</body>"
`;

exports[`CoreLibrary Unit Tests - MarkdownItRender Render Test: Styles are rendered (styleFilePaths) 1`] = `
"
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    <link rel="stylesheet" type="text/css" href="./test.css">

</head>
<body>
    <h1>test</h1>
<p>hello world</p>

</body>"
`;

exports[`CoreLibrary Unit Tests - MarkdownItRender Render Test: with logger 1`] = `
[MockFunction] {
  "calls": [
    [
      "render() called.",
    ],
    [
      "styles: %o",
      [
        "./test.css",
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
